### PR TITLE
Schnorr sign-to-contract and anti-exfil 

### PR DESCRIPTION
**Pull request by benma**
This adds sign-to-contract and anti-exfil protocol functionality to the schnorrsig module. It is based on the same functions that already exist for ECDSA here:

https://github.com/ElementsProject/secp256k1-zkp/blob/d22774e248c703a191049b78f8d04f37d6fcfa05/include/secp256k1_ecdsa_s2c.h

Some commits and functions were adapted from the original work here: https://github.com/bitcoin-core/secp256k1/pull/589/

### CODE REVIEW

File: src/modules/schnorrsig/main_impl.h [hunk 1]
```diff
@@ -11,6 +11,63 @@
 #include "../../../include/secp256k1_schnorrsig.h"
 #include "../../hash.h"
 
+static uint64_t s2c_opening_magic = 0x5d0520b8b7f2b168ULL;
+
+static const unsigned char s2c_data_tag[16] = "s2c/schnorr/data";
+static const unsigned char s2c_point_tag[17] = "s2c/schnorr/point";
```
Comment (benma) [inline src/modules/schnorrsig/main_impl.h:17]: If these tags are okay, I can add functions with precomputed states if desired.

**Comment by real-or-random** [comment]
Nice!

By the way, is there a reason why this uses S2C instead of simple addition (Scheme 6 in https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-March/017667.html) besides the fact that the ECDSA code uses S2C. Are you aware of any reason why the S2C variant would be preferable over Scheme 6? (I think @jonasnick once told me that he wasn't aware of all the various variants when the work was started.)    

---

It seems the tests fails on s390x, which is a big-endian machine, see https://cirrus-ci.com/task/6416044628639744?logs=cat_tests_log#L4  Is there some code in this PR where endianness could be a problem? I don't see any but I only had a brief look.

**Comment by benma** [comment]
> By the way, is there a reason why this uses S2C instead of simple addition

I did it this way for two reasons:

- sign-to-contract seems useful also for other uses, see https://github.com/ElementsProject/secp256k1-zkp/issues/154
- It seemed to be the easiest approach, seeing that the same was already implemented for ECDSA. That means it should need less time to code and review and maintain.

About s390x: I am puzzled, I don't see anything related to endianness in the code related to the failing test, but maybe I am missing something. I also replaced the uint64_t magic by an 8-byte buffer just in case there could be problems with large integers, but it didn't seem to be the cause.

**Comment by real-or-random** [comment]
> About s390x: I am puzzled, I don't see anything related to endianness in the code related to the failing test, but maybe I am missing something. I also replaced the uint64_t magic by an 8-byte buffer just in case there could be problems with large integers, but it didn't seem to be the cause.

According to the implementation of `counting_illegal_callback_fn` excepts an `int32_t` and not `size_t`. Maybe this is the problem.

File: include/secp256k1_schnorrsig.h [hunk 2]
```diff
@@ -13,6 +15,63 @@ extern "C" {
...
+ *  This structure is not opaque, but it is strongly discouraged to read or write to
+ *  it directly.
+ *
+ *  The exact representation of data inside is implementation defined and not
+ *  guaranteed to be portable between different platforms or versions. It can
+ *  be safely copied/moved.
+ */
+typedef struct {
+    /* magic is set during initialization */
+    uint64_t magic;
```
Comment (benma) [inline include/secp256k1_schnorrsig.h:31]: I took this for a commit from @jonasnick, but I am actually unsure what the purpose is of the magic. Absent programming errors inside the library, could something go wrong if there was no magic field? Are there guidelines of when to use a magic (most structs don't contain any)?

**Comment by benma** [comment]
> > About s390x: I am puzzled, I don't see anything related to endianness in the code related to the failing test, but maybe I am missing something. I also replaced the uint64_t magic by an 8-byte buffer just in case there could be problems with large integers, but it didn't seem to be the cause.
> 
> According to the implementation of `counting_illegal_callback_fn` excepts an `int32_t` and not `size_t`. Maybe this is the problem.

Great catch, that seemed to be the problem! Fixed now.

**Comment by benma** [comment]
Most of the commits are taken as-is from @jonasnick and @apoelstra, with the added exfil functions working the same way as in the ECDSA version in [secp256k1-zpk](https://github.com/ElementsProject/secp256k1-zkp/tree/master/src/modules/ecdsa_s2c).

It would be great if you could review this and let me know if there is any obstacles to getting this merged. I'd like to use this in the BitBox02 to extend the anti-exfil protocol to Schnorr sigs in Taproot inputs.

**Comment by real-or-random** [comment]
> It would be great if you could review this and let me know if there is any obstacles to getting this merged

I just want to let you know that this on our radar. Things can move slow here and I think most of the contributors have been pretty busy in the remaining weeks (and at least I will still be busy with other stuff for ~2 weeks). I can't speak for the others but this is a feature I'd like to see included in the library, so I'll certainly come back to this soon.

**Comment by benma** [comment]
Rebased to fix conflicts with `master`.

**Comment by benma** [comment]
@real-or-random @jonasnick @apoelstra any chance progress here? The code should straight forward, with a little bit of concentrated effort we could get this merged used in production :pray:

File: include/secp256k1_schnorrsig.h [hunk 3]
```diff
@@ -69,26 +128,36 @@ SECP256K1_API_VAR const secp256k1_nonce_function_hardened secp256k1_nonce_functi
...
+ *                a sign-to-contract commitment (i.e. the `s2c_data32` argument
+ *                is not NULL).
+ *    s2c_data32: pointer to a 32-byte data to create an optional
+ *                sign-to-contract commitment to if not NULL (can be NULL).
  */
 typedef struct {
     unsigned char magic[4];
     secp256k1_nonce_function_hardened noncefp;
     void *ndata;
+    secp256k1_schnorrsig_s2c_opening* s2c_opening;
```
[the remark this answers is not in the document]
Comment (benma) [reply include/secp256k1_schnorrsig.h:151]: >  I believe the library will access uninitialized fields. Is that right?

That is right, good catch.

There could be multiple solutions:

1. As you suggest, the magic can be used as a version field.
2. The s2c_opening struct itself contains a magic field. Instead of calling `secp256k1_schnorrsig_s2c_opening_init(s2c_opening)` to initialize it, we could have the caller do this instead and check that the magic is set properly. 
3. not adding these fields to extraparams but add another function secp256k1_schneorrsig_sign_s2c instead.

It seems to me that option 2 is worse, as if the opening magic is not set properly (e.g. by a user using the old headers with the new library), then the function would error. This is better than proceeding silently, but it still is incompatible and forces the user act. The alternative to erroring is to ignore these fields to maintain compatibility, but that would be strange for new library users that want to use the functionality and don't see a proper error message when they fail to initialize the opening struct.

Option 3 seems backwards as the point of `secp256k1_schnorrsig_extraparams` struct was, I presume, to be able to extend it in a backwards compatible fashion.

Option 2 probably makes sense to do for the same reason that the extraparams struct is initialized by the user - to prevent corrupting memory in case the user passes the wrong pointer, and to possibly use it as a version field in the future.  

TL;DR: I think defining a new magic in the extraparams struct is the best option, and we should also have the caller initialize the `secp256k1_schnorrsig_s2c_opening` struct with the magic. What do you think? 

As a sidenote: maybe it would be a good to introduce a version check or init function for this library in general, so that if breaking changes are necessary in the future, it could be caught early at runtime.

[the remark this answers is not in the document]
Comment (jonasnick) [reply include/secp256k1_schnorrsig.h:151]: The extraparams struct was built with the intention that `magic` acts as versioning for the struct and therefore allows backward-compatible changes (there's some mention of this in the original `schnorrsig_sign_custom` PR). I implemented a variant of the idea [here](https://github.com/jonasnick/versioned-param). I was imagining to change `schnorrsig_sign_custom` to something like this:

```
int secp256k1_schnorrsig_sign_custom(..., secp256k1_schnorrsig_extraparams *params_in) {
    secp256k1_schnorrsig_extraparams params;

    if (memcmp(params_in->magic, old_magic, sizeof(params_in->magic)) == 0) {
        secp256k1_extraparams_old *params_tmp;
        copy_old_into_new(&params, params_tmp);
    } else if (memcmp(params_in->magic, SECP256K1_SCHNORRSIG_EXTRAPARAMS_MAGIC, sizeof(params_in->magic)) == 0) {
        params = *params_in;
    }
    /* ... */
}
```

While this does work in practice, I'm not sure how compliant this is with the C standard. It would access the old extraparams struct through a pointer to a new extraparams struct. Padding shouldn't be an issue here because it only accesses the first member of the struct.

It would be better if there was a solution where correctness is more obvious.

[the remark this answers is not in the document]
Comment (real-or-random) [reply include/secp256k1_schnorrsig.h:151]: > I'm not sure how compliant this is with the C standard.

If you read the standard strictly, then it's just UB to have conflicting declarations for the same function: https://stackoverflow.com/a/69620952 

In practice, and probably even with a more pragmatic reading of the standard, it's unclear what should go wrong. Following your example code, the callee will read the magic value by accessing bytes through an lvalue of type unsigned char. That's always allowed (even if the new struct type has stricter alignment requirements than the old struct type). Moreover, the magic is guaranteed to be at the beginning of the struct, and it's explicitly allowed to use pointers to a struct as pointers to the first element (after an appropriate cast). After determining the version, the callee will read the struct members through an lvalue of the correct type (not identical type with same name but  "compatible type"), which is also allowed.

As the callee starts with reading characters, it's very much like passing an unsigned char* directly to the function and then casting it a pointer to the right struct type. I think we can do this in practice.

[the remark this answers is not in the document]
Comment (benma) [reply include/secp256k1_schnorrsig.h:160]: Thanks for the helpful inputs everyone!

I added a new commit `use the magic in the schnorrsig extraparams struct for versioning` that implements this for easier review. I can squash it in the end. Please take a look.

**Comment by benma** [comment]
@jonasnick @sipa @real-or-random should I continue here or open an equivalent PR in secp256k1-zkp, or something else?

It seems close to being ready for production but got stuck due to uncertainty about whether this belongs in this repo in the first place.

**Comment by jonasnick** [comment]
@benma I agree with sipa's comment above that a specification would be helpful. According to our [CONTRIBUTING.md](https://github.com/bitcoin-core/secp256k1/blob/master/CONTRIBUTING.md#adding-new-functionality-or-modules) it is recommended to provide a specification and security arguments in order to add a new module. I think the relevance criterion for the new module is met.

**Comment by zachariahevans174** [comment]
Thank you for all of your hard work and assistance on this project.

**Comment by zachariahevans174** [comment]
God Bless.!

File: include/secp256k1_schnorrsig.h [hunk 4]
```diff
@@ -13,6 +15,63 @@ extern "C" {
...
+     * verifying we could compute the EC commitment of original_pubnonce and the
+     * data and negate if this would not be a valid nonce. But this would prevent
+     * batch verification of sign-to-contract commitments. */
+    int nonce_is_negated;
+} secp256k1_schnorrsig_s2c_opening;
+
+/** The signer commitment in the anti-exfil protocol is the original public nonce. */
+typedef secp256k1_pubkey secp256k1_schnorrsig_anti_exfil_signer_commitment;
+
+/** Parse a sign-to-contract opening.
```
[the remark this answers is not in the document]
Comment (benma) [reply include/secp256k1_schnorrsig.h:44]: I added a specification in the header (in a new commit for easier review, will squash in the end).

The spec is mostly a copy/paste from

- https://github.com/BlockstreamResearch/secp256k1-zkp/blob/6152622613fdf1c5af6f31f74c427c4e9ee120ce/include/secp256k1_ecdsa_s2c.h#L100

A very similar spec was present in @jonasnick's original PR:

- https://github.com/bitcoin-core/secp256k1/pull/590/files#diff-75af67a78d848e432dcf1b975d64978c8d731530fcbde6c9b15edf7da556af00R65

**Comment by benma** [comment]
> @benma I agree with sipa's comment above that a specification would be helpful. According to our [CONTRIBUTING.md](https://github.com/bitcoin-core/secp256k1/blob/master/CONTRIBUTING.md#adding-new-functionality-or-modules) it is recommended to provide a specification and security arguments in order to add a new module. I think the relevance criterion for the new module is met.

Thanks. I added the specification in the comments, see https://github.com/bitcoin-core/secp256k1/pull/1140#discussion_r1796701547.

**Comment by benma** [comment]
Would be nice to move forward, I think it should be a small effort to finalize it from here. @jonasnick

**Review by jonasnick — COMMENTED** [review]
Sorry for the late reply @benma.

I don't think the text you added to the header qualifies as a specification. It's more like an API guide mixed in with some technical details. A defining property of a specification is that it's detailed enough to implement a compliant implementation. If someone were to create an alternative implementation, they would have to look at the libsecp256k1 code and they would have to create their own test vectors because the PR doesn't contain any. Also, the header also doesn't justify the design decision to go with scheme 6.

I think we should discuss what kind of specification we should have for anti-exfil and how to argue for it's security.

I also noticed that the anti-exfil protocol is not exhaustively tested for protocol failure. There only appears to be a test for successful completion under honest protocol participants. Maybe that would best go into an example. So I doubt that "it should be a small effort to finalize it from here". But as you know that almost never happens anyway :)

**Comment by benma** [comment]
@jonasnick haha, I thought this was close because it's at about the same level as the ECDSA variant in secp256k1-zkp :sweat_smile: 

I can work on better unit test coverage and a spec and test vectors, but I am not sure what "security arguments" entails. Could you give some examples, or for efficiency's sake, maybe even propose a draft?

**Comment by jonasnick** [comment]
In secp256k1-zkp we don't really have guidelines for merging new modules except that the maintainers need to be convinced. Here, we try to be more systematic. 

I haven't thought deeply about how a security argument would look like or if we need one. When I was looking at the PR 2 days ago I went through a phase of "this is obviously broken; the host can extract the signing key with a Wagner's attack; why did I ever think this was secure". Luckily that phase passed. However, that could be an indication that the security of the scheme is not trivial. This should be discussed with @sipa and @real-or-random as well.
